Create CORS docs (#519)
commerce-docs/microsite-commerce-storefront: src/content/docs/setup/configuration/cors-setup.mdx

---

import { Aside } from '@astrojs/starlight/components';
import Tasks from '@components/Tasks.astro';
import Task from '@components/Task.astro';
import Link from '@components/Link.astro';

Cross-Origin Resource Sharing (CORS) is a browser security feature that restricts web pages from making requests to a different domain than the one serving the page. CORS is required when your storefront runs on a different domain or port than your Adobe Commerce backend. 

<Aside type="caution" title="Production recommendation">
CORS should be a last resort for production. Prefer same-origin delivery of the storefront and API (for example, via Fastly VCL) to avoid CORS entirely.

But during development, it's a common scenario for storefronts to run on a different domain or port than the Adobe Commerce backend, so CORS is required. Without CORS, browsers will block cross-origin requests and you'll see errors like the following in the browser console:

```txt
Access to fetch at 'https://commerce.example.com/graphql' from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present
```
</Aside>

> During development, it's common for storefronts to run on a different domain or port than the Adobe Commerce backend. For example, if your storefront is running on `http://localhost:3000` and your Adobe Commerce backend is running on `https://commerce.example.com`, you will need to configure CORS to allow requests from the storefront to the backend.

Reviewer: This paragraph is duplicative of the one on line 16.

## Quick test: Is CORS working?

> Your Adobe Commerce GraphQL endpoint (typically `/graphql`) is the primary endpoint that requires CORS configuration. If you already have CORS configured, test it by replacing `YOUR_STOREFRONT_URL` with your actual frontend URL:

Reviewer: I don't think you can change the endpoint location.

```bash
curl -I -X OPTIONS https://your-commerce-backend.com/graphql \
  -H "Origin: YOUR_STOREFRONT_URL" \
  -H "Access-Control-Request-Method: POST"
```

Look for these headers in the response:
- `Access-Control-Allow-Origin: YOUR_STOREFRONT_URL` (or `*` if using a wildcard)
- `Access-Control-Allow-Methods: GET, POST, OPTIONS`

If you see these headers with your storefront URL, CORS is configured correctly. If not, follow the setup steps below.

## Quick steps

<Tasks>

<Task>

### Add CORS support

To add CORS headers to your Adobe Commerce GraphQL endpoints, you have several options:

- **Server configuration**: Configure CORS headers at the web server level (Nginx, Apache)
- **Custom module**: Implement a custom Adobe Commerce module to add CORS headers
- **Third-party module**: Use a community module such as [graycore/magento2-cors](https://github.com/graycoreio/magento2-cors) or similar

> Each approach has different implementation steps. If using a module-based approach:

Reviewer: I would not add installation steps for a 3rd party module.

Consider adding links for Nginx/Apache configuration. I found these

https://enable-cors.org/server_apache.html
https://enable-cors.org/server_nginx.html

Also, I would list custom module last. It's a lot harder than the other two options

1. Install via Composer:
   ```bash
   composer require <module-package-name>
   ```

2. Enable the module and run setup:
   ```bash
   bin/magento module:enable <Module_Name>
   bin/magento setup:upgrade
   bin/magento cache:flush
   ```

Refer to your chosen implementation's documentation for specific configuration steps.

</Task>

<Task>

### Configure CORS settings

Configure your CORS implementation to allow requests from your storefront domain(s) to your Adobe Commerce GraphQL endpoint.

Configuration methods vary by implementation:
- **Server configuration**: Consult your web server documentation (Nginx, Apache)
- **Custom module**: Follow your implementation's configuration approach

> - **Third-party modules**: Refer to the module's documentation (e.g., [graycore/magento2-cors configuration](https://github.com/graycoreio/magento2-cors#configuration))

Reviewer: ```suggestion
- **Third-party modules**: Refer to the module's documentation
```

If you see these headers with your storefront URL, CORS is configured correctly. If not, follow the setup steps below.

## When do you need CORS?

You need CORS when your storefront runs on a different domain or port than your Adobe Commerce backend (for example, `http://localhost:3000` → `https://commerce.example.com`).

Without CORS, browsers block cross-origin requests and you'll see errors like:

```txt
Access to fetch at 'https://commerce.example.com/graphql' from origin 'http://localhost:3000' 
has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present
```

## Quick steps

<Tasks>

<Task>

### Add CORS support

To add CORS headers to your Adobe Commerce GraphQL endpoints, you can configure your server, implement a custom module, or use a third-party CORS module (such as [graycore/magento2-cors](https://github.com/graycoreio/magento2-cors)).

If using a CORS module, install it via Composer:

```bash
composer require <module-package-name>
```

Refer to your chosen module's documentation for specific installation and configuration steps.

</Task>

<Task>

### Enable the module

If using a CORS module, enable it and run setup commands:

```bash

> php bin/magento module:enable <Module_Name>

Reviewer: We generally leave out the `php` from the command line. I've seen other pages in this repo that also omit it, but I don't know what's standard here.

Author: If it's a general practice to leave it off, i'll do the same here.

php bin/magento setup:upgrade
php bin/magento cache:flush
```

</Task>

<Task>

### Configure CORS in the Magento Admin

> The module adds a **CORS Whitelist** configuration section in the admin panel.

Reviewer: ```suggestion
The module adds a **CORS Whitelist** configuration section in the Admin.
```

> Navigate to **Stores → Configuration → General → Web → CORS Whitelist** and configure the following fields:

Reviewer: ```suggestion
Navigate to **Stores** > Configuration **General** > **Web** > CORS Whitelist** and configure the following fields:
```

Reviewer: I think you're talking about Graycore here, but you made the setup generic

composer require <module-package-name>
php bin/magento module:enable <Module_Name>

You don't know exactly where the generic module updates the Admin UI, and you definitely don't know the fields.

- `http://localhost:3000` ≠ `https://localhost:3000` (different protocol)
- `http://localhost:3000` ≠ `http://localhost:5173` (different port)
- `http://localhost:3000` ≠ `http://127.0.0.1:3000` (different hostname)
- No trailing slashes: `http://localhost:3000` not `http://localhost:3000/`

#### Allowed Methods

Set to: `GET,POST,OPTIONS`

#### Allowed Headers

Set to: `Content-Type,Authorization,X-Requested-With`

Add `Store` if using multistore.

#### Allow Credentials

Enable this if your storefront sends cookies or authentication credentials.

</Task>

<Task>

### Test your CORS configuration

Open your storefront in a browser and check the Network tab in DevTools:
- Look for preflight `OPTIONS` requests to your GraphQL endpoint
- Verify the response includes headers like:
  - `Access-Control-Allow-Origin: <your-storefront-url>` (or `*` if using a wildcard)
  - `Access-Control-Allow-Methods: GET, POST, OPTIONS`
  - `Access-Control-Allow-Headers: Content-Type, Authorization`

If configured correctly, your GraphQL requests should succeed without CORS errors.

</Task>

</Tasks>

## Common pitfalls

<Aside type="caution" title="Avoid these common mistakes">

> **Forgot to clear cache:** After changing CORS settings, always run `php bin/magento cache:flush`. A stale cache is the #1 cause of "it's not working" reports.

Reviewer: ```suggestion
**Forgetting to clear cache:** After changing CORS settings, always run `bin/magento cache:flush`. A stale cache is the primary cause of "it's not working" reports.
```

**Using `*` with credentials:** You cannot use a wildcard `*` origin when `Allow Credentials` is enabled. Browsers will block the request.

> **Missing `OPTIONS` method:** If you forget to include `OPTIONS` in Allowed Methods, all CORS requests will fail. All CORS requests start with an OPTIONS preflight request.

Reviewer: ```suggestion
**Missing the `OPTIONS` method:** All CORS requests start with an OPTIONS preflight request. If you forget to include `OPTIONS` in **Allowed Methods**, all CORS requests will fail.
```

---
title: CORS Setup
description: Quickstart guide to install and configure graycore/magento2-cors for local/dev use and provides guidance for Fastly-based production patterns.
---

import { Aside } from '@astrojs/starlight/components';
import Tasks from '@components/Tasks.astro';
import Task from '@components/Task.astro';
import Options from '@components/Options.astro';
import Option from '@components/Option.astro';
import Link from '@components/Link.astro';

<Aside type="caution" title="Production recommendation">
CORS should be a last resort for production. Prefer same-origin delivery of storefront and API (for example, via Fastly VCL) to avoid CORS entirely.
</Aside>

## Quick test: Is CORS working?

If you already have CORS configured, test it quickly:

```bash
curl -I -X OPTIONS https://your-commerce-backend.com/graphql \
  -H "Origin: http://localhost:3000" \
  -H "Access-Control-Request-Method: POST"
```

Look for these headers in the response:

> - `Access-Control-Allow-Origin: http://localhost:3000`

Reviewer 2: this is wrong - having localhost in ACAO does not imply correctness. Rather, they need to know what urls they will use in frontend, and then set that in ACAO. If more than one (ie `main--aem-boilerplate-commerce--hlxsites.aem.live` and `some-branch--aem-boilerplate-commerce--hlxsites.aem.live` and `localhost:3000`, I think the only option is to use wildcard ACAO, ie `*`.

- **Development**: Storefront at `http://localhost:3000` accessing Commerce at `https://commerce.example.com`
- **Staging**: Storefront at `https://staging-storefront.example.com` accessing Commerce at `https://commerce-staging.example.com`
- **Production**: Storefront at `https://storefront.example.com` accessing Commerce at `https://commerce.example.com`

Without CORS configured, browsers will block your storefront's GraphQL requests. You'll see an error in the browser console like:

```txt
Access to fetch at 'https://commerce.example.com/graphql' from origin 'http://localhost:3000' 
has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present
```

This is a standard CORS error enforced by all modern browsers (Chrome, Firefox, Safari, Edge).

## Quick steps

<Tasks>

<Task>

### Install the CORS module

Install the Graycore CORS package to your Adobe Commerce backend (v2.x required for Storefront):

```bash
composer require graycore/magento2-cors:^2.0

> ```

Reviewer 2: Are we really suggesting a third party module that we don't own? We should just say something like "add CORS headers to your GraphQL resolvers/endpoints....one of the ways you can do that might be with <module> or <cdn>", etc.

Reviewer 3: Agree

</Task>

<Task>

### Enable and configure the module

Enable the module and run upgrades:

```bash
php bin/magento module:enable Graycore_Cors
php bin/magento setup:upgrade
php bin/magento cache:flush
```

</Task>

<Task>

### Configure CORS in Magento Admin

The module adds a **CORS Whitelist** configuration section in the admin panel.

Navigate to **Stores → Configuration → Web → CORS Whitelist** and configure the following fields:

<Aside type="note" title="Locating CORS settings in admin">
In Commerce admin: **Stores** → **Configuration** → **Settings** → **Web** → **CORS Whitelist**

> </Aside>

Reviewer 2: Are you documenting this third party module now? Again - we should not need this - instead, I would just link to it as _an option_, and reference their docs.

Reviewer 3: We have same problem in Fastly: https://experienceleague.adobe.com/en/docs/commerce-on-cloud/user-guide/cdn/setup-fastly/fastly-configuration

But I agree, less content in our docs and just link them to GH repo

2. **Cloudflare Workers / CDN Edge Functions:**
   Use edge functions to route API requests to your backend dynamically.

3. **Reverse proxy (Nginx/Apache):**
   Configure your web server to proxy `/graphql` to the backend.

**Resources:**
- <Link href="https://developer.fastly.com/reference/vcl/" text="Fastly VCL documentation" />
- Consult your DevOps team or CDN provider for implementation

**Note:** This approach requires CDN/infrastructure expertise. If you don't have these resources, use CORS configuration (Option 2) instead.

</Option>

<Option>

### CORS configuration

If you must use different domains (e.g., `storefront.example.com` → `commerce.example.com`), use the CORS configuration above with specific allowed origins (never use `*` in production).

</Option>

</Options>

## More details

For troubleshooting, configuration examples, and debug steps, see the detailed reference: [CORS Troubleshooting](/setup/configuration/cors-troubleshooting/)

## References

> - <Link href="https://github.com/graycoreio/magento2-cors" text="graycore/magento2-cors" /> - Official module repository with configuration details

Reviewer 2: Official by whom? Do we own this? I don't feel comfortable documenting this as our recommendation.

---
title: CORS Setup
description: Guide to configuring CORS headers for your Adobe Commerce GraphQL endpoints to enable cross-origin requests from your storefront.
---

import { Aside } from '@astrojs/starlight/components';
import Tasks from '@components/Tasks.astro';
import Task from '@components/Task.astro';
import Options from '@components/Options.astro';
import Option from '@components/Option.astro';
import Link from '@components/Link.astro';

<Aside type="caution" title="Production recommendation">
CORS should be a last resort for production. Prefer same-origin delivery of the storefront and API (for example, via Fastly VCL) to avoid CORS entirely.
</Aside>

## Quick test: Is CORS working?

If you already have CORS configured, test it by replacing `YOUR_STOREFRONT_URL` with your actual frontend URL:

```bash

> curl -I -X OPTIONS https://your-commerce-backend.com/graphql \

Reviewer: The fact that the `graphql` endpoint is at issue should be explicitly mentioned before the sample command.

  -H "Origin: YOUR_STOREFRONT_URL" \
  -H "Access-Control-Request-Method: POST"
```

Look for these headers in the response:
- `Access-Control-Allow-Origin: YOUR_STOREFRONT_URL` (or `*` if using a wildcard)
- `Access-Control-Allow-Methods: GET, POST, OPTIONS`

If you see these headers with your storefront URL, CORS is configured correctly. If not, follow the setup steps below.

## When do you need CORS?

> Cross-Origin Resource Sharing (CORS) is a browser security feature that restricts web pages from making requests to a different domain than the one serving the page. CORS is required when your storefront runs on a different domain or port than your Adobe Commerce backend. For example:

Reviewer: This paragraph should be at the top, above the Caution. Maybe you can change the H2 to Use cases and salvage the rest of the section with minimal rewrite/reorg

- **Development**: Storefront at `http://localhost:3000` accessing Commerce at `https://commerce.example.com`
- **Staging**: Storefront at `https://staging-storefront.example.com` accessing Commerce at `https://commerce-staging.example.com`
- **Production**: Storefront at `https://storefront.example.com` accessing Commerce at `https://commerce.example.com`

Without CORS configured, browsers will block your storefront's GraphQL requests. You'll see an error in the browser console like:

```txt
Access to fetch at 'https://commerce.example.com/graphql' from origin 'http://localhost:3000' 
has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present
```

> This is a standard CORS error enforced by all modern browsers (Chrome, Firefox, Safari, Edge).

Reviewer: Just say all supported browsers instead of listing some of them
